pkg/controller/storageversiongc/gc_controller.go

```diff
@@ -0,0 +1,297 @@
+				continue
+			}
+			serverStorageVersions = append(serverStorageVersions, ssv)
+		}
+		if !hasStaleRecord {
+			continue
+		}
+		if err := c.updateOrDeleteStorageVersion(&sv, serverStorageVersions); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return utilerrors.NewAggregate(errors)
+}
+
+func (c *Controller) syncStorageVersion(name string) error {
+	sv, err := c.kubeclientset.InternalV1alpha1().StorageVersions().Get(context.TODO(), name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// The problematic storage version that was added/updated recently is gone.
+		// Nothing we need to do here.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	hasInvalidID := false
+	var serverStorageVersions []apiserverinternalv1alpha1.ServerStorageVersion
+	for _, v := range sv.Status.StorageVersions {
+		lease, err := c.kubeclientset.CoordinationV1().Leases(metav1.NamespaceSystem).Get(context.TODO(), v.APIServerID, metav1.GetOptions{})
+		if err != nil || lease == nil || lease.Labels == nil ||
+			lease.Labels[controlplane.IdentityLeaseComponentLabelKey] != controlplane.KubeAPIServer {
+			// We cannot find a corresponding identity lease from apiserver as well.
+			// We need to clean up this storage version.
+			hasInvalidID = true
+			continue
+		}
+		serverStorageVersions = append(serverStorageVersions, v)
+	}
+	if !hasInvalidID {
+		return nil
+	}
+	return c.updateOrDeleteStorageVersion(sv, serverStorageVersions)
+}
+
+func (c *Controller) onAddStorageVersion(obj interface{}) {
+	castObj := obj.(*apiserverinternalv1alpha1.StorageVersion)
+	c.enqueueStorageVersion(castObj)
+}
+
+func (c *Controller) onUpdateStorageVersion(oldObj, newObj interface{}) {
+	castNewObj := newObj.(*apiserverinternalv1alpha1.StorageVersion)
+	c.enqueueStorageVersion(castNewObj)
+}
+
+// enqueueStorageVersion enqueues the storage version if it has entry for invalid apiserver
+func (c *Controller) enqueueStorageVersion(obj *apiserverinternalv1alpha1.StorageVersion) {
+	for _, sv := range obj.Status.StorageVersions {
+		lease, err := c.leaseLister.Leases(metav1.NamespaceSystem).Get(sv.APIServerID)
+		if err != nil || lease == nil || lease.Labels == nil ||
+			lease.Labels[controlplane.IdentityLeaseComponentLabelKey] != controlplane.KubeAPIServer {
+			// we cannot find a corresponding identity lease in cache, enqueue the storageversion
+			klog.V(4).Infof("Observed storage version %s with invalid apiserver entry", obj.Name)
+			c.storageVersionQueue.Add(obj.Name)
+			return
+		}
+	}
+}
+
+func (c *Controller) onDeleteLease(obj interface{}) {
+	castObj, ok := obj.(*coordinationv1.Lease)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		castObj, ok = tombstone.Obj.(*coordinationv1.Lease)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a Lease %#v", obj))
```

Comment:
Should probably fix this in a followup to have cached the set of leases we care about, or to list them in this case. It should be rare.

Reply:
I need to verify if such a case corrupts the lease informer's cache.

If not, the storage version worker will eventually garbage collect any storage version that has non-existing lease when resync kicks in (default 12 hours), regardless if we get the lease delete event or not. 

It could be a problem if leases that we don't care about are triggering too much tombstone, and causing the GC to spam log where it shouldn't